test/test_linesearch.jl

```diff
@@ -63,4 +64,43 @@
     @test nbk > 0
     @test nbW > 0
   end
+
+  @testset "Don't allocate" begin
+    n = 200
+    nlp = SimpleModel(n)
+    x = nlp.meta.x0
+    g = zeros(n)
+    d = -40 * ones(n)
+    lm = LineModel(nlp, x, d)
+
+    al1 = @allocated obj(nlp, x)
+    al2 = @allocated obj(lm, 1.0)
+    @test al1 == al2
+
+    al1 = @allocated grad!(nlp, x, g)
+    al2 = @allocated grad!(lm, 1.0, g)
+    @test al1 + 16 == al2
+
+    al1 = @allocated objgrad!(nlp, x, g)
+    al2 = @allocated objgrad!(lm, 1.0, g)
+    al1 = @allocated objgrad!(nlp, x, g)
+    al2 = @allocated objgrad!(lm, 1.0, g)
+    @test al1 + 64 == al2
+
+    al1 = @allocated hprod!(nlp, x, d, g)
+    al2 = @allocated hess!(lm, 1.0, g)
+    @test al1 + 16 == al2
+
+    h₀ = obj(lm, 0.0)
+    slope = grad(lm, 0.0)
+    (t, gg, ht, nbk, nbW) = armijo_wolfe(lm, h₀, slope, g)
+    al = @allocated armijo_wolfe(lm, h₀, slope, g)
+    @test al == (8 + nbk) * 32 + 16
+
+    for bk_max = 0:8
+      (t, gg, ht, nbk, nbW) = armijo_wolfe(lm, h₀, slope, g, bk_max=bk_max)
+      al = @allocated armijo_wolfe(lm, h₀, slope, g, bk_max=bk_max)
+      @test al == min(12.5, 7 + 1.5nbk) * 32 + 208
```

Comment:
I’m trying to understand where `armijo_wolfe` allocates. Do you have a sense of that?

Reply:
I count:
- 6 scalar keyword arguments;
- Creation of 9 variables `nbk`, `nbW`, `ht`, `slope_t`, `h_goal`, `fact`, `ϵ`, `Armijo`, `good_grad`;
- One `obj` call per `nbk` (which should be 16 allocations);
- One `objgrad!` call (should be 64 allocations);

There are 4 iterations, and there are some scalar operations. I imagined most of these were actually no-ops, so I can't explain the allocations yet. Especially the extra `208` that only happens inside the `for`.